File: frontend/src/api/monitoring.ts
/** Tipos del panel de gerencia → Sistema: espejo de backend/app/schemas/monitoring.py.
 * Tiempos en milisegundos; fechas en hora local del negocio, sin zona. */

export type CheckStatus = "ok" | "warning" | "critical" | "unknown"
export type WindowHours = 1 | 24 | 168 | 720

export interface Check {
  key: string
  label: string
  status: CheckStatus
  detail: string
  latency_ms: number | null
  hint: string | null
}

export interface Alert {
  severity: "warning" | "critical"
  title: string
  detail: string
  hint: string | null
  source: string
}

export interface MonitoringStatus {
  checked_at: string
  overall: "ok" | "warning" | "critical"
  checks: Check[]
  alerts: Alert[]
  config: {
    llm_host: string
    llm_model: string
    embedding_model: string
    moderation_model: string
    pronunciation_model: string
    tts_provider: string
    llm_max_concurrency: number
    llm_max_retries: number
    api_key_configured: boolean
    environment: string
  }
}

export interface MonitoringMetrics {
  hours: WindowHours
  start: string
  end: string
  bucket_minutes: number
  llm: {
    attempts: number
    calls: number
    failed_calls: number
    retried_calls: number
    failed_attempts: number
    input_tokens: number
    output_tokens: number
    truncated: number
    chat_ok: number
    chat_p50_ms: number | null
    chat_p95_ms: number | null
    tutor_p50_ms: number | null
    tutor_p95_ms: number | null
    queue_p95_ms: number | null
    tts_chars: number
  }
  llm_series: {
    bucket: string
    attempts: number
    errors: number
    input_tokens: number
    output_tokens: number
    p50_ms: number | null
    p95_ms: number | null
  }[]
  llm_by_purpose: {
    purpose: string
    operation: string
    attempts: number
    calls: number
    errors: number
    p50_ms: number | null
    p95_ms: number | null
    avg_input_tokens: number | null
    avg_output_tokens: number | null
    total_tokens: number
    truncated: number
  }[]
  llm_by_model: {
    model: string
    operation: string
    attempts: number
    errors: number
    p50_ms: number | null
    p95_ms: number | null
    input_tokens: number
    output_tokens: number
  }[]
  http: { requests: number; server_errors: number; client_errors: number; p50_ms: number | null; p95_ms: number | null }
  http_series: { bucket: string; requests: number; server_errors: number; p95_ms: number | null }[]
  routes: {
    method: string
    route: string
    requests: number
    server_errors: number
    client_errors: number
    p50_ms: number | null
    p95_ms: number | null
    max_ms: number
  }[]
  client_errors: number
  /** Agrupados por fallo idéntico: `at` es la última vez, `first_at` la primera. */
  failures: {
    at: string
    first_at: string
    occurrences: number
    source: "llm" | "http" | "client"
    title: string
    message: string
    detail: string | null
  }[]
}

File: backend/app/schemas/monitoring.py
"""Respuestas del panel de gerencia → Sistema (routers/monitoring.py)."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class CheckOut(BaseModel):
    key: str
    label: str
    status: Literal["ok", "warning", "critical", "unknown"]
    detail: str
    latency_ms: int | None = None
    hint: str | None = None


class AlertOut(BaseModel):
    severity: Literal["warning", "critical"]
    title: str
    detail: str
    hint: str | None = None
    source: str


class ActiveConfig(BaseModel):
    """Lo que ESTE proceso está usando de verdad (no lo que dice el .env):
    justo lo que no cuadraba cuando el contenedor arrastraba variables viejas."""

    llm_host: str
    llm_model: str
    embedding_model: str
    moderation_model: str
    pronunciation_model: str
    tts_provider: str
    llm_max_concurrency: int
    llm_max_retries: int
    api_key_configured: bool
    environment: str


class StatusOut(BaseModel):
    checked_at: datetime
    overall: Literal["ok", "warning", "critical"]
    checks: list[CheckOut]
    alerts: list[AlertOut]
    config: ActiveConfig


class LLMSummary(BaseModel):
    attempts: int
    calls: int
    failed_calls: int
    retried_calls: int
    failed_attempts: int
    input_tokens: int
    output_tokens: int
    truncated: int
    chat_ok: int
    chat_p50_ms: float | None
    chat_p95_ms: float | None
    tutor_p50_ms: float | None
    tutor_p95_ms: float | None
    queue_p95_ms: float | None
    tts_chars: int


class LLMPoint(BaseModel):
    bucket: datetime
    attempts: int
    errors: int
    input_tokens: int
    output_tokens: int
    p50_ms: float | None
    p95_ms: float | None


class LLMByPurpose(BaseModel):
    purpose: str
    operation: str
    attempts: int
    calls: int
    errors: int
    p50_ms: float | None
    p95_ms: float | None
    avg_input_tokens: float | None
    avg_output_tokens: float | None
    total_tokens: int
    truncated: int


class LLMByModel(BaseModel):
    model: str
    operation: str
    attempts: int
    errors: int
    p50_ms: float | None
    p95_ms: float | None
    input_tokens: int
    output_tokens: int


class HTTPSummary(BaseModel):
    requests: int
    server_errors: int
    client_errors: int
    p50_ms: float | None
    p95_ms: float | None


class HTTPPoint(BaseModel):
    bucket: datetime
    requests: int
    server_errors: int
    p95_ms: float | None


class RouteStats(BaseModel):
    method: str
    route: str
    requests: int
    server_errors: int
    client_errors: int
    p50_ms: float | None
    p95_ms: float | None
    max_ms: int


class Failure(BaseModel):
    at: datetime  # la ocurrencia más reciente
    first_at: datetime
    occurrences: int
    source: Literal["llm", "http", "client"]
    title: str
    message: str
    detail: str | None


class MetricsOut(BaseModel):
    hours: int
    start: datetime
    end: datetime
    bucket_minutes: int
    llm: LLMSummary
    llm_series: list[LLMPoint]
    llm_by_purpose: list[LLMByPurpose]
    llm_by_model: list[LLMByModel]
    http: HTTPSummary
    http_series: list[HTTPPoint]
    routes: list[RouteStats]
    client_errors: int
    failures: list[Failure]


class ClientErrorIn(BaseModel):
    """Lo que manda el navegador. Todo con tope de longitud: es un
    endpoint público y no puede servir para meter megas en la BD."""

    kind: Literal["error", "unhandledrejection"]
    message: str = Field(max_length=1000)
    source: str | None = Field(None, max_length=300)
    path: str | None = Field(None, max_length=300)
    stack: str | None = Field(None, max_length=4000)
    in_app: bool = True
